🌱 Refactor testenv and remove unused ones

vincepri opened the pull request:
Signed-off-by: Vince Prignano <vincepri@vmware.com>

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
This PR makes it so we can reuse one envtest across all our packages.

/assign @detiber @fabriziopandini

benmoss commented:
Looks good to me

vincepri commented:
/milestone v0.3.7

fabriziopandini commented:
Nice cleanup! 
/approve

vincepri commented:
@fabriziopandini @benmoss pls lgtm :)

detiber commented on test/helpers/envtest.go:
```diff
@@ -76,31 +74,42 @@ func newTestEnvironment() *TestEnvironment {
-	if err != nil {
-		return env, err
-	}
+func NewTestEnvironment() *TestEnvironment {
```
Does this need to keep track of the number of times this is called, otherwise it looks like any call to `Stop()` may tear down the environment for everyone.

detiber replied:
Either that or the first envtest environment that is stood up could possibly be treated separately from others, and addtional envtest environments are created as this is called usiing the `UseExistingCluster` flag, which would prevent teardown of the controlplane when `Stop()` is called.

vincepri replied:
`UseExistingCluster` flag uses the local kubeconfig to call a cluster, that's not suitable in this case I think

fabriziopandini commented:
/lgtm

fabriziopandini commented:
/hold
just noticed the breaking change.

detiber commented:
/hold

detiber commented:
holding, in the middle of a review, I think there are potential issues outside of the api change.

vincepri commented:
The test/ package isn't under API contract

benmoss commented on controlplane/kubeadm/controllers/suite_test.go:
```diff
@@ -17,71 +17,43 @@ limitations under the License.
-
 func init() {
 	klog.InitFlags(nil)
-	klog.SetOutput(GinkgoWriter)
```
I think this is just if you use the Ginkgo CLI, I tried to understand this once but I never got to anything being clear.

detiber replied:
I believe it is needed to aggregate the output for a test (hiding output if `-v` isn't specified to only show when errors occur, displaying the output in the test failure summary, etc).

benmoss replied:
I still have no idea what the logic is here: https://gist.github.com/benmoss/eb5c7190abe9d242b7de9063b76fa1a2

vincepri commented:
/retitle 🌱 Refactor testenv and remove unused ones

Addressed all comments, I've opted to clean everything up and avoid finding a complex solution to reuse envtest across packages, few things I've noticed/discovered:
- A full test environment was running in places where it shouldn't have been.
- Go may run tests from different packages in parallel.
- Go runs tests in a separate binary / context, it's not possible to use the same global package.

Cleaning everything up resulted in faster local runs. In general, we should be mindful that each time we spin up a new test environment in a new package it might affect the overall runtime of tests.

vincepri commented:
/hold cancel

sedefsavas commented on controllers/remote/suite_test.go:
```diff
@@ -42,9 +38,6 @@ import (
 // http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
 
 func init() {
-	klog.InitFlags(nil)
```
Can we delete this init() too?

vincepri replied:
Yes, thanks!

vincepri commented on scripts/ci-test.sh:
```diff
@@ -24,9 +24,10 @@ cd "${REPO_ROOT}" || exit 1
 # shellcheck source=./hack/ensure-go.sh
 source "${REPO_ROOT}/hack/ensure-go.sh"
 
-make generate test
```
I've removed the generate pre-req from here, given that we're already checking it in `pr-verify` presubmit. This reduced the runtime for tests in CI from ~12 minutes to ~3 mins

vincepri commented:
/test pull-cluster-api-test

benmoss commented:
Doesn't look like this has much of an impact on my machine (8 cores, 32gb ram). I ran `make test` but with it modified to add `-count=1` to avoid `go test` caching.

```
== master

real    0m51.821s
user    5m7.616s
sys     0m31.552s

real    0m54.547s
user    5m13.911s
sys     0m32.933s

real    0m53.169s
user    5m4.569s
sys     0m31.689s

== reuse-testenv

real    0m48.349s
user    4m40.661s
sys     0m29.396s

real    0m53.823s
user    4m40.636s
sys     0m29.350s

real    0m51.977s
user    4m40.644s
sys     0m30.269s
```

benmoss commented:
still a good cleanup though! /approve

detiber commented on controlplane/kubeadm/controllers/controller_test.go:
```diff
@@ -60,7 +60,7 @@ var _ = Describe("KubeadmControlPlaneReconciler", func() {
 	Context("Reconcile a KubeadmControlPlane", func() {
 		It("should return error if owner cluster is missing", func() {
-			clusterName, clusterNamespace := "foo", "default"
+			clusterName, clusterNamespace := "foo-1", "default"
```
It seems like this should be using a generated namespace rather than "default", possibly one created in `BeforeEach` and deleted in `AfterEach`

detiber commented on controlplane/kubeadm/controllers/controller_test.go:
```diff
@@ -97,12 +97,12 @@ var _ = Describe("KubeadmControlPlaneReconciler", func() {
+			Eventually(func() error {
+				_, err := r.Reconcile(ctrl.Request{NamespacedName: util.ObjectKey(kcp)})
+				return err
+			}, 10*time.Second).Should(HaveOccurred())
```
Why the switch to eventually here? Would it be better to have an eventually on the deletion being fully completed and then retrying the reconcile call instead?

fabriziopandini commented:
/lgtm

detiber commented on test/helpers/envtest.go:
```diff
@@ -21,95 +21,106 @@ import (
 func (t *TestEnvironment) Stop() error {
-	close(t.doneMgr)
-	return t.Environment.Stop()
+	t.doneMgr <- struct{}{}
```
Does this actually stop the Manager? It looks like based on the docs that the manager is expected to stay running until the channel is closed: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/manager#Runnable
